### PR TITLE
feat: update builtin catalog server

### PR DESCRIPTION
**Pull request by alexcodelf**
IMPORTANT: Please do not create a Pull Request without creating an issue first.

Problem:
Add builtin catalog server and record telemetry

Solution:
Support fetch git repositories from catalog server

Related Issue:

### CODE REVIEW

File: pkg/catalog/catalog.go [hunk 1]
```diff
@@ -14,48 +16,33 @@ import (
...
-			break
-		}
-	}
-
-	return list, nil
-}
-
 // GetRepos returns org and a list of repositories from the given catalog.
-func GetRepos(ctx context.Context, c *model.Catalog) ([]*scm.Repository, error) {
+func GetRepos(ctx context.Context, mc model.ClientSet, c *model.Catalog) ([]*vcs.Repository, error) {
```
Comment (gitlawr) [inline pkg/catalog/catalog.go:28]: make it unexported

File: pkg/catalog/catalog.go [hunk 2]
```diff
@@ -223,3 +210,38 @@ func (cs catalogSyncer) Do(_ context.Context, busMessage catalog.BusMessage) err
 
 	return nil
 }
+
+func GetRedirectURL(rawURL, userAgent string) (string, error) {
```
Comment (gitlawr) [inline pkg/catalog/catalog.go:214]: ditto

File: pkg/catalog/catalog.go [hunk 3]
```diff
@@ -223,3 +210,38 @@ func (cs catalogSyncer) Do(_ context.Context, busMessage catalog.BusMessage) err
...
+			TLSClientConfig: &tls.Config{
+				MinVersion: tls.VersionTLS12,
+			},
+		},
+		CheckRedirect: func(req *http.Request, via []*http.Request) error {
+			return http.ErrUseLastResponse
+		},
+	}
+
+	req, err := http.NewRequest(http.MethodGet, rawURL, nil)
```
Comment (gitlawr) [inline pkg/catalog/catalog.go:228]: Should we do http HEAD

File: pkg/vcs/repo.go [hunk 4]
```diff
@@ -200,3 +202,47 @@ func GetRepoRef(r *git.Repository, name string) (*plumbing.Reference, error) {
...
+		if opts.Page == 0 && opts.URL == "" {
+			break
+		}
+	}
+
+	return list, nil
+}
+
+// GetOrgFromGitURL parses the organization from the given URL.
+func GetOrgFromGitURL(source string) (string, error) {
```
Comment (gitlawr) [inline pkg/vcs/repo.go:236]: Does it work if it's a multi-part path, like gitlab subgroup?

Comment (alexcodelf) [reply pkg/vcs/repo.go:236]: yes， the sub group is tested `https://gitlab.com/terraform-seal-modules/sub`

File: pkg/vcs/client.go [hunk 5]
```diff
@@ -38,13 +40,28 @@ func NewClient(conn *model.Connector) (*scm.Client, error) {
...
 	case github.Driver:
-		return github.NewClientFromURL(url, token)
+		return github.NewClientFromURL(server, token)
 	case gitlab.Driver:
-		return gitlab.NewClientFromURL(url, token)
+		return gitlab.NewClientFromURL(server, token)
 	}
 
 	return nil, fmt.Errorf("unsupported SCM driver %q", driver)
 }
```
Comment (thxCode) [inline pkg/vcs/client.go:59]: can we return a scm.Client with User-Agent directly?

Comment (alexcodelf) [reply pkg/vcs/client.go]: updated

**Review by thxCode — APPROVED** [review]
LGTM

**Review by gitlawr — APPROVED** [review]
LGTM

**Pull request merged**